warehouse/attestations/services.py

```diff
@@ -0,0 +1,249 @@
+        )
+
+    # Empty attestation sets are not permitted; users should omit `attestations`
+    # entirely to upload without attestations.
+    if not attestations:
+        raise AttestationUploadError(
+            "Malformed attestations: an empty attestation set is not permitted"
+        )
+
+    # This is a temporary constraint; multiple attestations per file will
+    # be supported in the future.
+    if len(attestations) > 1:
+        metrics.increment("warehouse.upload.attestations.failed_multiple_attestations")
+
+        raise AttestationUploadError(
+            "Only a single attestation per file is supported",
+        )
+
+    return attestations
+
+
+@implementer(IIntegrityService)
+class NullIntegrityService:
+    def __init__(self, session):
+        warnings.warn(
+            "NullIntegrityService is intended only for use in development, "
+            "you should not use it in production due to the lack of actual "
+            "attestation verification.",
+            InsecureIntegrityServiceWarning,
+        )
+        self.db = session
+
+    @classmethod
+    def create_service(cls, _context, request):
+        return cls(session=request.db)
+
+    def parse_attestations(
+        self, request: Request, _distribution: Distribution
+    ) -> list[Attestation]:
+        publisher: OIDCPublisher | None = request.oidc_publisher
+        if not publisher or not publisher.publisher_name == "GitHub":
+            raise AttestationUploadError(
+                "Attestations are only supported when using Trusted "
+                "Publishing with GitHub Actions.",
+            )
+        return _extract_attestations_from_request(request)
+
+    def build_provenance(
+        self, request: Request, file: File, attestations: list[Attestation]
+    ) -> DatabaseProvenance:
+        return _build_provenance(request, file, attestations)
+
+
+@implementer(IIntegrityService)
+class IntegrityService:
+    def __init__(self, metrics: IMetricsService, session):
+        self.metrics: IMetricsService = metrics
+        self.db = session
+
+    @classmethod
+    def create_service(cls, _context, request):
+        return cls(
+            metrics=request.find_service(IMetricsService),
+            session=request.db,
+        )
+
+    def parse_attestations(
+        self, request: Request, distribution: Distribution
+    ) -> list[Attestation]:
+        """
+        Process any attestations included in a file upload request
+
+        Attestations, if present, will be parsed and verified against the uploaded
+        artifact. Attestations are only allowed when uploading via a Trusted
+        Publisher, because a Trusted Publisher provides the identity that will be
+        used to verify the attestations.
+        Only GitHub Actions Trusted Publishers are supported.
+        """
+        publisher: OIDCPublisher | None = request.oidc_publisher
+        if not publisher or not publisher.publisher_name == "GitHub":
```

Comment:
We should avoid special-casing Github here to make this easier to add new publishers in the future, something like this would be nice:

```suggestion
        if not publisher or not publisher.supports_attestations
```

Reply:
Good catch! Adding.

Reply:
Done! I've added a blanket-`False` `OIDCPublisher.supports_attestations` that can be overridden by each publisher type.